### PR TITLE
[AI-129] Capture all conversations in Cosmos DB for evaluation

### PR DESCRIPTION
**Pull request by roberthunterjr**
## Summary

- Adds `conversation-capture-store.js` — a Cosmos DB store that records full conversation records (user message, bot response, thread history, LLM metadata, citations) behind `CAPTURE_ALL_CONVERSATIONS`.
- Updates `callLLM` to return `{ metadata, botText, systemPromptVersion }` so listeners can persist full response text plus prompt-version metadata.
- Wires `captureConversation` into `app_mention.js` and `assistant/message.js` after successful LLM completion/finalization.
- Provisions a dedicated `conversations` Cosmos container in Bicep and aligns conversation retention to **180 days** (`defaultTtl: 15552000`, app-side `ttl: 15_552_000`).

## Security / Compliance updates in this PR

- Missing-field warning logs in conversation capture now include only missing field names (not raw user/channel/thread/message values).
- Conversation capture rejects `COSMOS_ENDPOINT + COSMOS_KEY` auth in `DEPLOYMENT_TYPE=production` and logs a warning to require connection string or managed identity for production.

## Configuration

| Env var | Default | Purpose |
|---|---|---|
| `CAPTURE_ALL_CONVERSATIONS` | `false` | Set to `true` to enable conversation capture |
| `COSMOS_CONVERSATIONS_CONTAINER` | `conversations` | Container name override |
| `SYSTEM_PROMPT_VERSION` | `v1` | Version tag persisted with captured records |

## Test coverage highlights

- Conversation capture retry/reconnect behavior covered (`429` retry, non-retryable no-retry, max-attempt exhaustion, reconnect path on `410`).
- Disabled-flag path covered for both unset and `'false'` values.
- Listener tests updated to verify `systemPromptVersion` propagation into capture payloads.
- `callLLM` return-shape tests updated for `systemPromptVersion`.

## Follow-up scope (tracked separately)

Hardening and refactor follow-up items are tracked in: **#57**

### CODE REVIEW

**Comment by roberthunterjr** [comment]
Tested in Insiders and ready for final review.

**Comment by roberthunterjr** [comment]
Visual evidence of the thread history and sources populating in insiders.

image

**Comment by stephenfuqua** [comment]
# PR #53 Review — [AI-129] Capture All Conversations in Cosmos DB for Evaluation
> Issue AI-129: Capture all conversations in Cosmos DB for evaluation
> Reviewed by: Security · Functionality · Maintainability · Usability · Test Coverage

---

## 🔴 Critical

**[Test Coverage]** Retry loop is completely untested — no test verifies that a retryable status code (410, 429, 449, 500, 503) triggers a second upsert attempt
Detail: The retry loop in `conversation-capture-store.js` is the most operationally critical behavior in the new module. The existing error test only verifies a generic `Error` throw, which hits the non-retryable path and returns on the first failure. The retryable path, backoff delay, and `maxAttempts` exhaustion are entirely untested.
Fix: Add three tests: (1) mock `upsert` to reject once with `{ statusCode: 429 }` then succeed — assert `mockUpsert` called twice; (2) mock with `{ statusCode: 400 }` — assert called exactly once (non-retryable = no retry); (3) mock `maxAttempts` rejections with retryable code — assert warning fires and upsert called `maxAttempts` times.

**[Test Coverage]** Reconnect logic (`container = null` reset on 410/503) is completely untested
Detail: When `RECONNECT_CODES` (410 or 503) are received, `container` is set to `null` so `getContainer` reconnects on the next attempt. If this regresses, all subsequent retries after a stale-session error would silently use a dead container reference. There is no test that validates `container` is cleared and a new client is instantiated.
Fix: Add a test that rejects attempt 1 with `{ statusCode: 410 }` and succeeds on attempt 2, then asserts `MockCosmosClient` (or `mockDatabase.container`) was called a second time, confirming reconnection.

---

## 🟠 High

**[Security]** PII (userId, channelId, threadTs, messageTs) logged in plaintext on the missing-fields warning path
Detail: In `conversation-capture-store.js`, when required fields are missing, the code logs `JSON.stringify({ userId, channelId, threadTs, messageTs, entryPoint })`. Slack user IDs are PII under many privacy frameworks. Depending on log aggregation config, these may be stored in plaintext infrastructure accessible to more people than Cosmos DB.
Fix: Log only which field *names* are missing, not their values: `const missing = [...].filter(k => !fields[k]); logger?.warn('Missing required fields: ' + missing.join(', '));`

**[Security]** `COSMOS_KEY` is accepted as a plain env var granting full account-level access with no production guard
Detail: A raw Cosmos DB account key grants full read/write/delete access to the entire account, not just the conversations container. If it leaks via log misconfiguration or a process.env dump in a crash report, all containers (feedback, interactions, conversations) are exposed.
Fix: Enforce managed identity as the only auth path in production. If `COSMOS_KEY` must remain for local dev, add a guard (`if (DEPLOYMENT_TYPE === 'local')`) and document that it must never be set in production.

**[Security · Functionality]** userMessage, botResponse, and threadHistory stored verbatim with no size cap; Cosmos has a 2 MB item limit
Detail: Full conversation text is written to Cosmos without length validation. This is high-volume (every conversation, not just explicit feedback), increasing the probability of hitting the limit. There is also no per-user opt-out mechanism for a feature that captures full message content.
Fix: Enforce a max character length on `userMessage` and `botResponse` (e.g., 50,000 chars) with a truncation marker and warn-log when triggered. Cap `threadHistory` entry count. Document the data-handling policy.

**[Maintainability]** Retry/reconnect logic duplicated verbatim across all three Cosmos store files — a future bug fix must be applied in three places
Detail: `formatCosmosError`, `toNumericCode`, `isEmulatorTarget`, `getRetryPolicy`, `getDelayMs`, `RETRYABLE_CODES`, `RECONNECT_CODES`, and the entire retry loop body are identical in `feedback-store.js`, `interaction-store.js`, and `conversation-capture-store.js`. This PR adds the third copy. `interaction-store.js` is already drifting from `feedback-store.js` in `getContainer` behavior and default database name (`fiona` vs. `chatbot`).
Fix: Extract a `cosmos-retry.js` or `cosmos-utils.js` shared module exporting the retry utilities and a `withCosmosRetry(fn, { logger, context })` wrapper. Each store retains only its container setup and document shape.

**[Maintainability]** `getContainer` client-construction logic also duplicated with already-visible per-store drift
Detail: The three-branch credential selection (`CONNECTION_STRING` → `ENDPOINT+KEY` → `ENDPOINT`+AAD) is copy-pasted in all three stores. `feedback-store.js` and `interaction-store.js` already differ in whether `DEPLOYMENT_TYPE` is read at module-load vs. inline, and in `warnedMissingConfig` handling.
Fix: Extract a `createCosmosClient({ connectionString, endpoint, key })` factory into the shared utils module.

**[Test Coverage]** `CAPTURE_ALL_CONVERSATIONS` is not tested with the string value `'false'` — only `undefined` is tested in the disabled block
Detail: The source uses `=== 'true'`. A common deployment error is setting `CAPTURE_ALL_CONVERSATIONS=false` (the string `'false'`), which the code correctly treats as disabled, but the test suite does not cover this case.
Fix: Add a test in the disabled describe block that sets `process.env.CAPTURE_ALL_CONVERSATIONS = 'false'` and asserts no upsert occurs.

**[Test Coverage]** The `COSMOS_ENDPOINT + COSMOS_KEY` and `COSMOS_ENDPOINT`-only (AAD/managed identity) auth paths are untested — only `COSMOS_CONNECTION_STRING` path is covered
Detail: The AAD path is the production auth path in Azure. An authentication regression on the managed-identity branch would go undetected.
Fix: Add a describe block that sets `COSMOS_ENDPOINT` + `COSMOS_KEY` (no connection string) and another with only `COSMOS_ENDPOINT`, asserting successful upsert in each.

**[Test Coverage]** `captureConversation` is not tested for the path where `streamer.stop()` throws — in that case capture is silently skipped
Detail: In both `app_mention.js` and `message.js`, `captureConversation` is called after `await streamer.stop(...)`. If `streamer.stop` throws, capture never runs. There is no assertion that this failure mode is expected and acceptable, making it a silent data loss path in the evaluation corpus.
Fix: Add a listener test that mocks `streamer.stop` to reject and asserts `captureConversation` was not called (to document the behavior as intentional) or restructure to call capture before `streamer.stop`.

---

## 🟡 Medium

**[Security · Functionality]** TTL is 360 days but the issue specifies 180 days for compliance — no documented justification for the deviation
Detail: `CONVERSATION_TTL_SECONDS = 31_104_000` (360 days) in `conversation-capture-store.js` and `defaultTtl: 31104000` in the Bicep. The issue explicitly flagged the TTL value for review as a compliance matter. The PR description notes "360 days" but provides no rationale for doubling the stated requirement.
Fix: Align to 180 days (`15_552_000`) per the issue, or add an explicit inline comment in both the code and Bicep explaining the business/legal rationale with sign-off documented.

**[Functionality]** System prompt version is absent from the captured document despite being explicitly listed in the requirements
Detail: The issue lists "LLM provider, system prompt version, citations/sources" as required metadata. The document captures `llmProvider`, `llmModel`, and `sources` but has no `systemPromptVersion` field. `llm-caller.js` has no concept of a system prompt version identifier.
Fix: Export a `SYSTEM_PROMPT_VERSION` constant from `llm-caller.js` (e.g., a short hash of `DEFAULT_SYSTEM_PROMPT` or a manually bumped semver), pass it through `callLLM`'s return value, and include it in the `captureConversation` call and stored document.

**[Usability]** Slack users receive no indication that their conversations are recorded for human evaluation
Detail: Unlike explicit feedback (thumbs up/down), this capture is passive and invisible to users. Depending on organizational policy and jurisdiction, passive retention of full conversation content for human review may require disclosure.
Fix: Add a brief notice in the assistant thread started handler or `/fiona help` output informing users that conversations may be used for quality evaluation.

**[Usability]** No success-path logging — operators enabling `CAPTURE_ALL_CONVERSATIONS=true` have no positive signal that capture is working
Detail: The `warnedMissingConfig` flag means even the "not configured" warning fires only once per process lifetime. After startup, a misconfigured deployment (wrong container name, wrong database) produces no logs at all. There is no `info` or `debug` log on a successful write.
Fix: Emit at least `logger.debug()` on each successful capture (e.g., `"Captured conversation: ${doc.id}"`). Consider a startup-time summary log: `"Conversation capture: enabled (container: conversations)"` or `"Conversation capture: disabled"`.

**[Maintainability]** `getRetryPolicy()` re-reads `process.env.NODE_ENV` on every call rather than at module initialization — inconsistent with every other env var in the file
Detail: If `NODE_ENV` changes mid-execution due to test setup/teardown side effects, the retry policy becomes non-deterministic. All other env vars are captured once at module load.
Fix: Compute `const RETRY_POLICY = Object.freeze(computeRetryPolicy())` once at module initialization instead of calling `getRetryPolicy()` at each retry loop entry.

**[Test Coverage]** `callLLM` success-path return shape `{ metadata, botText }` is not tested in the unit test suite — only error paths are covered in `llm-caller.aggregate-perplexity.test.js`
Detail: The listener tests mock `callLLM` entirely. No test calls the real `callLLM` successfully and asserts the return shape, so a regression that drops `botText` from the return value would not be caught at the unit level.
Fix: Add a `callLLM` success-path test that asserts `result.metadata` is an object and `result.botText` is a string.

**[Test Coverage]** Listener tests do not assert `threadHistory` or `sources` fields passed to `captureConversation`
Detail: These are the highest-value fields for evaluation purposes. A regression that passes `undefined` or the wrong array would go undetected in the listener tests.
Fix: In both `app-mention.test.js` and `message.test.js` capture tests, add assertions on `captureConversation`'s `threadHistory` and `sources` arguments.

**[Test Coverage]** `threadHistory` non-array normalization (`Array.isArray(threadHistory) ? threadHistory : []`) is untested
Detail: If a caller passes `null` or a non-array, the stored document should contain `[]`. This is not covered.
Fix: Add a test passing `threadHistory: null` and asserting `doc.threadHistory` equals `[]`.

---

## 🔵 Low

**[Security]** Document `id` (`${userId}_${threadTs}_${messageTs}`) does not include `teamId` — cross-workspace collision possible with `upsert` semantics
Detail: Slack user IDs are scoped per workspace but the `id` does not include `teamId`. Two users in different workspaces could share the same user ID and thread timestamps, causing silent document overwrite.
Fix: Include `teamId` in the document `id`: `${teamId}_${userId}_${threadTs}_${messageTs}`.

**[Functionality]** `captureConversation` is silently skipped when `botText` is an empty string (LLM produced no content)
Detail: `botText` is initialized to `''` and only set when `callPerplexityChat` returns text. An empty-response document in the evaluation corpus has no analytical value.
Fix: Add a guard in the listener (or in `captureConversation`) to skip capture when `botResponse` is falsy.

**[Maintainability]** `toNumericCode` relies on fragile prose string matching as a fallback for Cosmos SDK error codes
Detail: `if (message.includes('service is currently unavailable')) return 503` will silently stop working if the SDK changes its error message wording.
Fix: Add a `logger.warn()` when the string-fallback path is used, creating an observable signal if the SDK message changes. Add a comment citing the SDK version range this was tested against.

**[Maintainability]** `LLM_MODEL` export name is provider-neutral while the underlying constant is `PERPLEXITY_API_MODEL` — the export name disconnects the reader from the source
Fix: Either export under the original name or rename the internal constant to `LLM_MODEL` throughout.

**[Usability]** No error message when Cosmos is configured but the conversations container does not exist — first write silently fails with a generic Cosmos 404 warn
Fix: In `getContainer`, emit a warn that includes the expected container name when the container reference fails on first use: `"Conversations container 'conversations' not found — run Bicep deployment to provision it."`

---

## ⚪ Nit

**[Security]** Module-level singleton `container` reset on `RECONNECT_CODES` is a concurrent-request race condition — two simultaneous 503s both reset to `null` and both attempt reconnection
Fix: Use a "pending reconnect" promise to serialize reconnection attempts.

**[Maintainability]** `getDelayMs` has redundant nested `Math.floor` calls — inner floor is a no-op
Fix: `const jitter = Math.floor(Math.random() * Math.ceil(baseDelay * 0.2));`

**[Usability]** The function and variable names (`captureConversation`, `CAPTURE_ALL_CONVERSATIONS`) use "capture" where the rest of the codebase uses "record" (`recordFeedback`, `recordInteraction`)
Fix: Align to `recordConversation` / `RECORD_CONVERSATIONS` for consistency (or document the deliberate distinction).

---

## Specialist Verdicts

| Reviewer | Verdict |
|----------|---------|
| Security | Significant concerns: 360-day TTL doubles the stated compliance limit, PII leaks into warning logs, and `COSMOS_KEY` auth has no production guard — resolve before shipping to production. |
| Functionality | PARTIAL — Core wiring is correct and graceful degradation works, but the TTL deviates from the compliance spec without justification and system prompt version is entirely absent from the captured schema. |
| Maintainability | The three-way verbatim duplication of Cosmos retry infrastructure is the single most important issue to resolve; the duplicated boilerplate will compound future maintenance costs. |
| Usability | The operator/administrator experience has significant gaps: no positive success signal in logs, no startup-time feature-state summary, and no guidance when Cosmos is configured but the container is missing. |
| Test Coverage | WEAK — Happy path, disabled-flag no-op, and listener call-site integration are tested, but the retry loop and reconnect logic — the most complex new behavior in the module — have zero test coverage. |

---

## Summary

The core wiring is solid — zero overhead when disabled, graceful degradation on missing Cosmos config, correct entry-point coverage — but three issues should be resolved before merge: **(1)** the TTL is 360 days against a 180-day compliance requirement with no documented justification; **(2)** retry/reconnect logic in `conversation-capture-store.js` has zero test coverage for its most critical paths; and **(3)** the verbatim duplication of Cosmos retry infrastructure across three store files creates a maintenance liability that should be extracted to a shared module. The missing system prompt version field and PII-in-warnings issue are also worth addressing before the feature goes to production.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)


[AI-129]: https://edfi.atlassian.net/browse/AI-129?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: apps/fiona-slack/scripts/setup-cosmos-emulator.js [hunk 1]
```diff
@@ -31,17 +30,16 @@ try {
 
 const DATABASE_NAME = process.env.COSMOS_DATABASE || 'chatbot';
 const FEEDBACK_CONTAINER = process.env.COSMOS_CONTAINER || 'feedback';
-const INTERACTIONS_CONTAINER = 'interactions';
+const INTERACTIONS_CONTAINER = process.env.COSMOS_INTERACTIONS_CONTAINER || 'interactions';
+const CONVERSATIONS_CONTAINER = process.env.COSMOS_CONVERSATIONS_CONTAINER || 'conversations';
 
 // Build CosmosClient from connection string if available, otherwise fall back
 // to the well-known emulator endpoint + key.
-const tlsAgent = new https.Agent({ rejectUnauthorized: false });
```
Comment (stephenfuqua) [inline apps/fiona-slack/scripts/setup-cosmos-emulator.js:38]: Why was this removed? Don't we need this to connect to the self-signed certificate used by the Cosmos DB emulator?

File: apps/fiona-slack/src/agent/conversation-capture-store.js [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+function getRetryPolicy() {
+  if (process.env.NODE_ENV === 'test') {
+    return { maxAttempts: 2, baseDelayMs: 1, maxDelayMs: 5 };
+  }
+  if (isEmulatorTarget()) {
+    return { maxAttempts: 2, baseDelayMs: 200, maxDelayMs: 1000 };
+  }
+  return { maxAttempts: 3, baseDelayMs: 150, maxDelayMs: 800 };
+}
```
Comment (stephenfuqua) [inline apps/fiona-slack/src/agent/conversation-capture-store.js:80]: Interesting that the retry policy differs in this way. Seems reasonable. We should look across the entire Fiona application to make sure we are using retries consistently.

Just had Claude do a scan. Observations logged to: https://edfi.atlassian.net/browse/AI-138

No action required right now.

File: apps/fiona-slack/src/agent/conversation-capture-store.js [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+      }
+      return null;
+    }
+    client = new CosmosClient({ endpoint: COSMOS_ENDPOINT, key: COSMOS_KEY });
+  } else if (COSMOS_ENDPOINT) {
+    client = new CosmosClient({
+      endpoint: COSMOS_ENDPOINT,
+      aadCredentials: new DefaultAzureCredential(),
+    });
+  } else {
```
Comment (stephenfuqua) [inline apps/fiona-slack/src/agent/conversation-capture-store.js:114]: Do we really need all of these different authentication techniques? For example, if we can't use `COSMOS_KEY` in production, and the emulator is providing an entier connection string, then what is the value in having code to support use of endpoint & key?

In production are we using `COSMOS_ENDPOINT` with Azure credentials, or are we using the connection string?

File: apps/fiona-slack/src/agent/feedback-store.js [hunk 4]
```diff
@@ -6,30 +6,97 @@
...
-  if (connectionString) {
-    client = new CosmosClient({ connectionString });
-  } else if (endpoint && key) {
-    client = new CosmosClient({ endpoint, key });
-  } else if (endpoint) {
-    client = new CosmosClient({ endpoint, aadCredentials: new DefaultAzureCredential() });
+  if (COSMOS_CONNECTION_STRING) {
+    client = new CosmosClient(COSMOS_CONNECTION_STRING);
+  } else if (COSMOS_ENDPOINT && COSMOS_KEY) {
+    client = new CosmosClient({ endpoint: COSMOS_ENDPOINT, key: COSMOS_KEY });
```
Comment (stephenfuqua) [inline apps/fiona-slack/src/agent/feedback-store.js:94]: This connection logic is different than what we see in the `conversation-store`. Time to extract a shared module for generating a new client.

No action required right now. Created https://edfi.atlassian.net/browse/AI-139 to address this.

File: apps/fiona-slack/scripts/setup-cosmos-emulator.js (hunk 1 above)
Comment (stephenfuqua) [reply apps/fiona-slack/scripts/setup-cosmos-emulator.js:38]: We discussed in person: `NODE_TLS_REJECT_UNAUTHORIZED=0` is sufficient, not need for that code above.

I've tested and confirmed this.

**Review by stephenfuqua — APPROVED** [review]

File: apps/fiona-slack/src/agent/conversation-capture-store.js (hunk 3 above)
Comment (roberthunterjr) [reply apps/fiona-slack/src/agent/conversation-capture-store.js:114]: Good point. We are using the connection string in production. I know Default Azure is a fallback for container execution environments or when you log in using a CLI. Probably fine to remove, especially in this conversation capture context.
